rfcs/RFC Feature - USD in the Scene Pipeline.md

```diff
@@ -0,0 +1,545 @@
+
+_Below: RefExample.usda references HelloWorld.usda and overrides world's color._
+
+_HelloWorld.usda_
+```
+#usda 1.0
+(
+    defaultPrim = "hello"
+)
+
+def Xform "hello"
+{
+    double3 xformOp:translate = (4, 5, 6)
+    uniform token[] xformOpOrder = ["xformOp:translate"]
+
+    def Sphere "world"
+    {
+        float3[] extent = [(-2, -2, -2), (2, 2, 2)]
+        color3f[] primvars:displayColor = [(0, 0, 1)]
+        double radius = 2
+    }
+}
+```
+
+_RefExample.usda_
+```
+#usda 1.0
+over "refSphere2" (
+    prepend references = @./HelloWorld.usda@)
+    {
+        over "world"
+        {
+            color3f[] primvars:displayColor = [(1, 0, 0)]
+        }
+    }
+```
+  
+
+### Beyond the Basics
+
+I recommend reading through the USD tutorials here [https://graphics.pixar.com/usd/release/tut\_usd\_tutorials.html](https://graphics.pixar.com/usd/release/tut_usd_tutorials.html), and experimenting in Blender and Nvidia USD Composer with creating or importing scenes and exporting to usda format files.
+
+Note: Blender does not support references:  
+_"Like the USD exporter, the importer does not yet handle more advanced USD concepts, such as layers and references." [Universal Scene Description - Blender 4.2 Manual](https://docs.blender.org/manual/en/latest/files/import_export/usd.html)_
+
+Digital content creation (DCC) tools which takes advantage of USD references are Nvidia Isaac Sim, Nvidia USD Composer, and Pixar tooling.  
+
+Recommended Solution
+====================
+
+Solution Overview
+-----------------
+
+Build support for USD into the scene pipeline.
+
+Do this in three phases:
+
+1.  Upgrade O3DE to use Assimp's latest implementation
+    1.  Input: single USD geometry files that don't use layers or references
+    2.  Output: azmodel containing the USDs geometry 
+2.  Contribute back to Assimp and update scene builder to load USD files with references. Stages are flattened by adding in reference geometry. O3DE then processes USD like FBX and other scene file formats.
+    1.  Input: USD files which contains a hierarchy of references to other geometry inside of other USD files 
+    2.  Output: Flattened azmodel containing the geometry from the multiple USD files inside one azmodel file
+3.  Contribute back to Assimp giving API option to flatten stages or keep reference geom separate. Update O3DE to create a procedural prefab that mimics the relationship graph of USD files by matching models and transforms. This could be expanded later to allow O3DE developers to translate USD properties into O3DE component properties as well as add advanced shaders/materials.
+    1.  Input: USD files which contains a hierarchy of references to other geometry inside of other USD files
+    2.  Output: Procedural prefabs generated per USD. Root prefab preserves the hierarchy of the original USD.
```

Comment:
I think that its a good way to start phasing this in, but might actually be in conflict with both the purpose of stages and references, as well as a possible long term usage.  
Files referencing other files are used for a number of reasons,
1. Just to arrange things more neatly or split work apart between different people.  The final scene is intended to be re-assembled as a whole by loading everything.
2. To present alternative "variants" of a Thing in the scene, for the purposes of alleviating CPU or GPU or Memory load.  For example, there might be an Oak Tree object in the scene, but there are 2 available variants, a full detailed render-ready oak tree with a billion verts, or a billboard oak tree to switch it out for, when teh author is working on something other than rendering and would like their DCC tool to remain responsive.  In this case, the 2 variants are never expected to ever render at the same time or exist in the same scene at the same time.
3. To duplicate objects without duplicating data.  In this case, the same object is repeated in the scene over and over, but its the same object, so each one just references an existing USD but specifies a new transform for it.  
4. To provide a procedural object library - an example being a "chair" USD which references 20 different variants of a chair, but only loads a selected one.  This allows an artist to populate their scene with 100 chairs, but then choose a different variant for each such chair instance, to make them look unique, as well as from a data organizing perspective, allow other artists to add more available variants to the "chair" USD over time, expanding the library of avilable chair variants to use without invalidating existing scenes or data.

The basic approach of "flattening" every mesh it finds into one azmodel file can only actually potentially cover the first option, and does not cover the other options at all.  It might be a good start though, depending on how its written.

Its also probably why ASSIMP and others may not have tackled the problem - its not a binary problem, and "what parts of this USD do you want to realize into the scene" is not solvable by automation or brute force, its the choice of the given artist at the time of the given usage, and the combinatorics can explode too quickly to do a "for each variant, for each other variant, export each combination" situation...

for example,

car.usd pseudocode
```
  (root xform)
     VARIANT A:  "car_low_detail_fake_geometry_for_preview.usd"
     VARIANT B: 
        xform "car"
            VARIANT 1:  "lambo.usd"
            VARIANT 2:  "audi.usd"
       .. hundeds of cars, you can use in a parking garage
```

Workflow in your pixar DCC tool is to duplicate the car over and over in the parking garage and just click on each instance and choose a variant like "audi" or "lambo".   The world can also substitute all of them for variant A at an individual or global level to drop detail to nothing on the cars to save CPU and GPU and MEM.  Switching them back to Variant B will restore all the cars to full render detail.

from above, you can see that simply exporting everything into one atom model would be a mistake and would not be useful to anyone.  Especially if someone then has ParkingGarage.USD with hundreds of instances of the car.

As for longer term, There's only a couple ways I think we can actually support what is essentially the superpower of USD (points 2, 3, and 4 above).  

One way would be to make it so that the scene processing pipeline is actually reactive to the usage of the USD in the scene.  What this means is, there's no merging (or little marging, but no merging across variants) so in the above example, the most that gets exported is an atom mesh seperately for car.usd, lambo.usd, audi.usd, and car_low_detail_fake_geometry_for_preview.usd.

This would essentially mean no changes to assimp, those files would just be extracted into their own USDs and would not attempt to follow references, just flattening everything into the same USD into one.  

But the magic would be when it generates the procprefab for car, it would make components on the transforms that represented the objects - when it encounters a USD variant that would switch which USD it actually used, it could actually generate a component that allowed the user to switch the procprefab (output by the pipeline) was selected as a child, and switch between the available variants that way when editing individual entities.  (if not that, then at least switch between available flattened atom meshes that represent sub-usds)

A further optimization would be those switcher components being allowed to, when edited, also edit the associated scene export settings, so that variants unused are simply not present or compiled (and recompile and become usable when selected).

Whats interesting about that, is that the functionality for having a component or a property of the Prefab Component (which is the component on an entity which stores the reference to the sub prefab to spawn the heirarhcy of as a child to itself) is not a USD *thing*.  Adding this functionality would support USD variants, but also allow our prefabs to use that functionality without USD being involved, too.

Any thoughts here?